File: leetcode_problems/typescript/src/lc1238_circular_permutation_in_binary_representation.ts
function circularPermutation(n: number, start: number): number[] {
  function is_valid(cur: number, prev: number): boolean {
    let xr = cur ^ prev;
    return xr > 0 && (xr & (xr - 1)) == 0;
  }

  function gen_neighbors(num: number, n: number): number[] {
    let nb = [];
    for (let i = 0; i < n; i++) {
      nb.push(num ^ (1 << i));
    }
    return nb;
  }

  function backtrack(
    i: number,
    k: number,
    stack: number[],
    used: boolean[],
    nb: Map<number, number[]>
  ): boolean {
    if (i >= k) {
      return stack[k - 1] !== -1 && is_valid(stack[0], stack[stack.length - 1]);
    }

    let prev = stack[i - 1];
    let cnb = nb.get(prev);
    if (!cnb) {
      return false;
    }

    for (let j of cnb) {
      if (used[j] || !is_valid(j, stack[i - 1])) {
        continue;
      }

      used[j] = true;
      stack[i] = j;
      if (backtrack(i + 1, k, stack, used, nb)) {
        return true;
      }
      stack[i] = -1;
      used[j] = false;
    }
    return false;
  }

  let k = 2 ** n;
  let used = new Array(k).fill(false);
  used[start] = true;

  let stack = new Array(k).fill(-1);
  stack[0] = start;

  let nb = new Map<number, number[]>();
  for (let i = 0; i < k; i++) {
    nb.set(i, gen_neighbors(i, n));
  }

  backtrack(1, k, stack, used, nb);
  return stack;
}

File: leetcode_problems/cpp/lc1238_circular_permutation_in_binary_representation.cpp
#include <vector>
#include <cmath>
#include <unordered_map>
using std::pow;
using std::unordered_map;
using std::vector;

class Solution
{
public:
    bool is_valid(int cur, int prev)
    {
        int xr = cur ^ prev;
        return xr > 0 && (xr & (xr - 1)) == 0;
    }

    bool backtrack(int i, int k, vector<int> &stack, bool *used, unordered_map<int, vector<int>> &nb)
    {
        if (i >= k)
        {
            return is_valid(stack[0], stack[stack.size() - 1]);
        }

        for (int j : nb[stack[i - 1]])
        {
            if (used[j] || !is_valid(j, stack[i - 1]))
            {
                continue;
            }

            used[j] = true;
            stack[i] = j;
            if (backtrack(i + 1, k, stack, used, nb))
            {
                return true;
            }
            stack[i] = -1;
            used[j] = false;
        }

        return false;
    }

    vector<int> gen_neighbors(int num, int n)
    {
        vector<int> nb;
        for (int i = 0; i < n; i++)
        {
            nb.push_back(num ^ (1 << i));
        }
        return nb;
    }

    vector<int> circularPermutation(int n, int start)
    {
        int k = pow(2, n);
        bool *used = new bool[k];
        for (int i = 0; i < k; i++)
        {
            used[i] = false;
        }
        used[start] = true;

        vector<int> stack(k);
        for (int i = 0; i < stack.size(); i++)
        {
            stack[i] = -1;
        }
        stack[0] = start;

        unordered_map<int, vector<int>> nb;
        for (int i = 0; i < k; i++)
        {
            nb[i] = gen_neighbors(i, n);
        }

        backtrack(1, k, stack, used, nb);
        delete[] used;
        return stack;
    }
};
